web_src/js/svg.js

```diff
@@ -120,6 +119,45 @@ export const SvgIcon = {
     className: {type: String, default: ''},
   },
   render() {
-    return h('span', {innerHTML: svg(this.name, this.size, this.className)});
+    const svgStr = svgs[this.name];
+    if (!svgStr) throw new Error(`Unknown SVG icon: ${this.name}`);
+
+    // parse the SVG string to 2 parts
+    // * svgInnerHtml: the inner part of the SVG, will be used as the content of the <svg> VNode
+    // * svgOuter: the outer part of the SVG, including attributes
+    // the builtin SVG contents are clean, so it's safe to use `indexOf` to split the content:
+    // eg: <svg outer-attributes>${svgInnerHtml}</svg>
+    const p1 = svgStr.indexOf('>'), p2 = svgStr.lastIndexOf('<');
+    if (p1 === -1 || p2 === -1) throw new Error(`Invalid SVG icon: ${this.name}`);
+    const svgInnerHtml = svgStr.slice(p1 + 1, p2);
+    const svgOuterHtml = svgStr.slice(0, p1 + 1) + svgStr.slice(p2);
+    const svgDoc = parser.parseFromString(svgOuterHtml, 'image/svg+xml');
+    const svgOuter = svgDoc.firstChild;
+
+    // https://vuejs.org/guide/extras/render-function.html#creating-vnodes
+    // the `^` is used for attr, set SVG attributes like 'width', `aria-hidden`, `viewBox`, etc
+    const attrs = {};
+    for (const attr of svgOuter.attributes) {
+      if (attr.name === 'class') continue;
+      attrs[`^${attr.name}`] = attr.value;
+    }
+    attrs[`^width`] = this.size;
+    attrs[`^height`] = this.size;
```

Comment:
What's the `^` for?

Reply:
There is a comment above.

Reply:
Hmm I still don't get it, that link does not explain it. Why can we have `id` or `class` but `^height` needs the `^`?

Reply:
`.` is for `prop`, eg: `disabled` / `textContent` -> `patchDOMProp`

`^` is for attribute, eg: `data-xxx=xxx`, `width=xxx` -> `patchAttr`

The code is here

https://github.com/vuejs/core/blob/c6eb3ccccee8e43e7aafcbc3c9ededecc565fdf0/packages/runtime-dom/src/patchProp.ts

It's a little difficult to understand.

Reply:
These prefixes are used to avoid conflicting. Without these prefixes, Vue would "guess" what it is.

With these prefixes, it's explicitly defined that these are attributes.

Reply:
Understood. Vue is even worse then React when it comes to mangling DOM attribute names, but ultimately it's a failure of the DOM standard to even have a difference between a prop and an attr.